### ISSUE TITLE
Implement the same user invite flow (join links) for groups, that we have for topics

### ISSUE THREAD

**Issue by loorm** [citizenos/citizenos-fe-old#325]
**What is the problem?**
Invite works differently for topics and groups. Join links and QR codes are available for topics, but not for groups.

**Why is this a problem?**
Options for inviting people to join groups are very limited. But groups are one of the main organising features on Citizen OS and especially when public groups will be implemented, it's important to be able to invite many people to join in. For example, for Telliskivi Creative City group, it would be very difficult to add everyone by single e-mail copy-pastes. Much easier to throw the join link into FB group Telliskivi asukad and have everyone click it.

**Possible solution.**
Bring the same SHARE solution into groups, as we have in topics (with links and QR-codes).

![group_invite_link](https://user-images.githubusercontent.com/22661743/67959412-60295880-fc01-11e9-92f0-cae8e5226e7e.jpg)

**Comment by loorm** [comment]
Triage 15: we will implement the new topic invite functionality for groups too, not the old one. New one is in #311 (complete with UI images).

**Comment by loorm** [comment]
Move forward with this, once #311 is done.

**Comment by anettlinno** [comment]
Triage 45. Using the same design as Kevin provided in Issue #311. Sending to development.

**Comment by ilmartyrk** [comment]
Tasklist:
- [ ] Update Group model
- [ ] Update tests
- [ ] add FE design

Administrative tasks:
- [ ] Update CHANGELOG files
- [ ] Inform marketing of new feature(s)
- [ ] Inform localization on new translations

**Comment by KatiVellak** [comment]
Legally reviewed and discussed, no impact.

**Comment by tiblu** [comment]
## TODO

### The Main Quest - Share/join links 
- [x] New Models/tables for join links. - **DONE:** GroupJoin
- [x]  Update API `POST /api/groups` - token generation for fresh Group
- [x] New API `POST /api/groups/:groupId/join` to (re)generate join link for a Group
- [x]  Update API `GET /api/users/:userId/groups/:groupId` and others that should give out GroupJoin info.
- [x] ~~`GET /api/v2/search` returning of join token for Group~~ Skip
- [x] API DOCS: Update docs, seems manual work right now :( - **DONE:** TBH the docs seems to contain invalid syntax and info, specially when it comes to "security" part. Just added my endpoints there, postboned fixing all of it as we need to find a better way of generating docs so that it is up to date.
- [x] `POST /api/groups/join/:token` to join a Topic
    - [x]  Make sure Joining with a link DOES NOT OVERWRITE existing permissions
- [x] Activities in all of the API-s!
    - [x] `POST /api/groups/join/:token` - User joined a Group - **DONE: no change**
    - [x] `PUT /api/users/:userId/groups/:groupId/join` - User generated a NEW join token WITH level
    - [x] `PUT /api/users/:userId/groups/:groupId/join/:token` - User set a new level for an EXISTING token
- [x] `API GET /activities` STOP RETURNING CONTEXT?
- [x] DB migration:
    - [x] CREATE TABLE "GroupJoins" 
    - [x] Generate "GroupJoin" for every existing Group with default level - READ
- [x] Update UX
    - [x] Assets
    - [x] Dropdown
    - [x] UI updates (layout etc)
    - [x] Warning that regenerating link will invalidate the previous links
    - [x] Mobile friendlyness
    - [x] Activities!
- [ ] BUG: FE - registration flow brakes join flow - after registration and login you are not redirected back to /join

### Side quests

- [x] Social Share
- [x] QR Code download as PDF **NOT DONE:** Separated to another task - https://github.com/citizenos/citizenos-fe/issues/832
- [x] Slug support. **NOT DONE:** I feel like it does not offer much.

### Administrative tasks

- [x] Update CHANGELOG files
- [ ] Inform marketing of new feature(s)
- [ ] Inform localization on new translations

**Comment by tiblu** [comment]
## API released 29.10.2021

* DB backup before release - B2152

**Comment by tiblu** [comment]
## Feature released 05.11.2021

![image](https://user-images.githubusercontent.com/7405383/140502350-2c21ef2c-a9f5-4205-8f44-5dc54a7bdc07.png)

Implementation notes:

* "SHARE" tab is only visible to Group admins.
* The "SHARE" tab can ONLY be accessed for the "My Groups" view by clicking "invite" link on the "Users" tab and then switching to "SHARE" tab.
* When non-authenticated User opens a `/join` link, they are redirected to login. IF the User chooses to register with username/password, after registration, User is not sent back to the `/join` link. IF the same User chooses to log in, User will be redirected to the `/join` link, thus getting access to the Group.

**Comment by anettlinno** [comment]
Tested and works well. Really nice addition!
